modules/core/src/main/java/com/illposed/osc/transport/udp/OSCPortIn.java

```diff
@@ -59,73 +62,131 @@
 			final OSCParserFactory parserFactory,
-			final OSCPacketDispatcher dispatcher,
+			final List<OSCPacketListener> packetListeners,
 			final SocketAddress local,
 			final SocketAddress remote)
 			throws IOException
 	{
 		super(local, remote);
 
-		this.listening = false;
-		this.daemonListener = true;
-		this.resilient = false;
-		this.parserFactory = parserFactory;
-		if (dispatcher == null) {
-			this.dispatcher = new OSCPacketDispatcher();
-			// HACK We do this, even though it is against the OSC (1.0) specification,
-			//   because this is how it worked in this library until Feb. 2015.,
-			//   and thus users of this library expect this behaviour by default.
-			this.dispatcher.setAlwaysDispatchingImmediately(true);
-		} else {
-			this.dispatcher = dispatcher;
-		}
+		this.listening       = false;
+		this.daemonListener  = true;
+		this.resilient       = false;
+		this.parserFactory   = parserFactory;
+		this.packetListeners = packetListeners;
 	}
 
 	public OSCPortIn(
 			final OSCParserFactory parserFactory,
-			final OSCPacketDispatcher dispatcher,
+			final List<OSCPacketListener> packetListeners,
 			final SocketAddress local)
 			throws IOException
 	{
-		this(parserFactory, dispatcher, local, new InetSocketAddress(
-				OSCPort.generateWildcard(local), 0));
+		this(
+			parserFactory,
+			packetListeners,
+			local,
+			new InetSocketAddress(OSCPort.generateWildcard(local), 0)
+		);
 	}
 
-	public OSCPortIn(final OSCParserFactory parserFactory, final SocketAddress local)
-			throws IOException
+	private static List<OSCPacketListener> packetListeners(
+		final boolean defaultListener
+	)
+	{
+		List<OSCPacketListener> listeners = new ArrayList<OSCPacketListener>();
+
+		if (defaultListener) {
+			OSCPacketDispatcher dispatcher = new OSCPacketDispatcher();
+			// HACK We do this, even though it is against the OSC (1.0) specification,
+			//   because this is how it worked in this library until Feb. 2015.,
+			//   and thus users of this library expect this behaviour by default.
+			dispatcher.setAlwaysDispatchingImmediately(true);
+
+			listeners.add(dispatcher);
+		}
+
+		// If one is opting out of using the default listener, one is expected to
+		// use `addPacketListener` to provide at least one listener to handle
+		// packets. Failure to do this will result in unspecified behavior.
+
+		return listeners;
+	}
+
+	public OSCPortIn(
+		final OSCParserFactory parserFactory, final SocketAddress local,
+		final boolean defaultListener
+	) throws IOException
 	{
-		this(parserFactory, null, local);
+		this(parserFactory, packetListeners(defaultListener), local);
 	}
 
 	/**
 	 * Creates an OSC-Port that listens on the given local socket.
 	 * @param local address to listen on
+	 * @param defaultListener whether to use the default packet listener, which
+	 * unwraps bundles and dispatches individual messages to message listeners
+	 * @throws IOException if we fail to bind a channel to the local address
+	 */
+	public OSCPortIn(
```

Comment:
here and in other places, please use the formatting like in the rest of the lib:


		public OSCPortIn(
				final OSCParserFactory parserFactory,
				final OSCPacketDispatcher dispatcher,
				final SocketAddress local)
				throws IOException
		{